arch/x86/core/intel64/locore.S

```diff
@@ -0,0 +1,599 @@
+
+irq:
+	pushq %rsi
+	movq $_kernel, %rsi
+
+	/*
+	 * Bump the IRQ nesting count and move to the next IRQ stack.
+	 * That's sufficient to safely re-enable interrupts, so if we
+	 * haven't reached the maximum nesting depth yet, do it.
+	 */
+
+	incl _kernel_offset_to_nested(%rsi)
+	addq $CONFIG_ISR_STACK_SIZE, ist1
+	cmpl $CONFIG_ISR_DEPTH, _kernel_offset_to_nested(%rsi)
+	jz 1f
+	sti
+1:	cmpl $1, _kernel_offset_to_nested(%rsi)
+	je irq_enter_unnested
+
+	/*
+	 * if we're a nested interrupt, we have to dump the state to the
+	 * stack. we play some games here to re-arrange the stack thusly:
+	 *
+	 * SS RSP RFLAGS CS RIP RAX RSI RCX RDX RBX
+	 * RDI RBP R8 R9 R10 R11 R12 R13 R14 R15
+	 * X86_FXSAVE_SIZE bytes of SSE data <-- RSP points here
+	 *
+	 * note that the final value of RSP must be 16-byte aligned here,
+	 * both to satisfy FXSAVE/FXRSTOR but also to honor the C ABI.
+	 */
+
+irq_enter_nested: /* Nested IRQ: dump register state to stack. */
+	pushq %rcx
+	movq 16(%rsp), %rcx /* RCX = vector */
+	movq %rax, 16(%rsp) /* looks like we pushed RAX, not the vector */
+	pushq %rdx
+	pushq %rbx
+	pushq %rdi
+	pushq %rbp
+	pushq %r8
+	pushq %r9
+	pushq %r10
+	pushq %r11
+	pushq %r12
+	pushq %r13
+	pushq %r14
+	pushq %r15
+	subq $X86_FXSAVE_SIZE, %rsp
+	fxsave (%rsp)
+	jmp irq_dispatch
+
+irq_enter_unnested: /* Not nested: dump state to thread struct for __resume */
+	movq _kernel_offset_to_current(%rsi), %rsi
+	andb $(~_THREAD_SWAPPED), _thread_offset_to_thread_state(%rsi)
+	fxsave _thread_offset_to_sse(%rsi)
+	movq %rbx, _thread_offset_to_rbx(%rsi)
+	movq %rbp, _thread_offset_to_rbp(%rsi)
+	movq %r12, _thread_offset_to_r12(%rsi)
+	movq %r13, _thread_offset_to_r13(%rsi)
+	movq %r14, _thread_offset_to_r14(%rsi)
+	movq %r15, _thread_offset_to_r15(%rsi)
+	movq %rax, _thread_offset_to_rax(%rsi)
+	movq %rcx, _thread_offset_to_rcx(%rsi)
+	movq %rdx, _thread_offset_to_rdx(%rsi)
+	movq %rdi, _thread_offset_to_rdi(%rsi)
+	movq %r8, _thread_offset_to_r8(%rsi)
+	movq %r9, _thread_offset_to_r9(%rsi)
+	movq %r10, _thread_offset_to_r10(%rsi)
+	movq %r11, _thread_offset_to_r11(%rsi)
+	popq %rax /* RSI */
+	movq %rax, _thread_offset_to_rsi(%rsi)
+	popq %rcx /* vector number */
+	popq %rax /* RIP */
+	movq %rax, _thread_offset_to_rip(%rsi)
+	popq %rax /* CS: discard */
+	popq %rax /* RFLAGS */
+	movq %rax, _thread_offset_to_rflags(%rsi)
+	popq %rax /* RSP */
+	movq %rax, _thread_offset_to_rsp(%rsi)
+	popq %rax /* SS: discard */
+
+irq_dispatch:
+	movq x86_irq_funcs(,%rcx,8), %rbx
+	movq x86_irq_args(,%rcx,8), %rdi
+	call *%rbx
+
+#ifdef CONFIG_X2APIC
+	call z_x2apic_eoi
+#else
+	xorl %eax, %eax
+	movl %eax, (CONFIG_LOAPIC_BASE_ADDRESS + LOAPIC_EOI)
+#endif
+
+	movq $_kernel, %rsi
+	cli
+	subq $CONFIG_ISR_STACK_SIZE, ist1
+	decl _kernel_offset_to_nested(%rsi)
+	/* not nested, exit via __resume (might change threads) */
+#ifdef CONFIG_STACK_SENTINEL
+	call z_check_stack_sentinel
+#endif
+	jz __resume
+
+irq_exit_nested:
+	fxrstor (%rsp)
```

Comment:
Similarly, you have two separate assembly blocks for "switch to"

Reply:
Yeah, the mechanism is necessarily different when leaving a nested IRQ than it is under other circumstances... so there are two separate assembly blocks.